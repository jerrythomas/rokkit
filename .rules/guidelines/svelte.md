# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
