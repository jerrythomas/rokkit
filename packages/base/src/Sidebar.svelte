<script>
	import Menu from './Menu.svelte'

	export let items
</script>

<aside
	class="bg-accent-200 text-accent-500 p-8 h-full w-80 space-y-6 flex-shrink-0"
>
	<slot name="header" />
	<Menu {items} />
	<slot name="footer" />
</aside>
