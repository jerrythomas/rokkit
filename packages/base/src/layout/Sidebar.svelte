<script>
	import Menu from './Menu.svelte'

	export let items
	export let activePage
</script>

<aside class="sidebar">
	<slot name="header" />
	<Menu {items} {activePage} />
	<slot name="footer" />
</aside>

<style lang="postcss">
	aside {
		@apply flex flex-col p-8 h-full w-80 flex-shrink-0;
	}
</style>
