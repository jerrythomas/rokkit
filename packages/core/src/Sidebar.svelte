<aside class="sidebar flex flex-col">
	<slot name="header" />
	<slot />
	<slot name="footer" />
</aside>
