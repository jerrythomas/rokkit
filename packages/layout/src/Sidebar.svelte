<aside class="flex flex-col sidebar">
	<slot name="header" />
	<slot />
	<slot name="footer" />
</aside>
