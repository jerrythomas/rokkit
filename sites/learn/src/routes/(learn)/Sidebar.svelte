<script>
	import { getContext } from 'svelte'
	import { Icon } from '@rokkit/core'
	let site = getContext('site')
	/**
	 * @type {HTMLElement}
	 */
	let sidebar
	let search

	function toggle(site) {
		if (site.sidebar) {
			sidebar.classList.remove('-translate-x-full')
		} else {
			sidebar.classList.add('-translate-x-full')
		}
	}
	$: if (sidebar) toggle($site)
</script>

<aside
	bind:this={sidebar}
	class="flex flex-col gap-4 bg-skin-base h-full flex-shrink-0 w-full md:w-1/2 lg:w-80 bg-skin-base absolute top-0 left-0 -translate-x-full transform transition duration-200 ease-in-out z-5 border-r border-skin-subtle overflow-auto"
	class:lg:relative={$site.sidebar}
>
	<nav
		class="flex h-10 items-center w-full gap-2 border-b border-b-skin-subtle"
	>
		<Icon
			name="i-rokkit:action-cross"
			role="button"
			class="border-r border-r-skin-subtle"
			on:click={() => ($site.sidebar = false)}
		/>
		<input
			type="search"
			placeholder="search"
			bind:value={search}
			class="embedded"
		/>
	</nav>
	<slot />
</aside>
