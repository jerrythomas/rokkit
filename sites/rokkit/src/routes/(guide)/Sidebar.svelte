<script>
	import { getContext } from 'svelte'
	let site = getContext('site')
	/**
	 * @type {HTMLElement}
	 */
	let sidebar

	function toggle(site) {
		if (site.sidebar) {
			sidebar.classList.remove('-translate-x-full')
		} else {
			sidebar.classList.add('-translate-x-full')
		}
	}
	$: if (sidebar) toggle($site)
</script>

<aside
	bind:this={sidebar}
	class="flex flex-col flex-shrink-0 bg-skin-50 w-80 p-4 absolute inset-y-0 left-0 -translate-x-full transform transition duration-200 ease-in-out z-10 border-r border-skin-base"
	class:lg:relative={$site.sidebar}
>
	<slot />
</aside>
