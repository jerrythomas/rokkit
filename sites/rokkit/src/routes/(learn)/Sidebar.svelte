<script>
	import { getContext } from 'svelte'
	// import { Icon } from '@rokkit/core'
	let site = getContext('site')
	/**
	 * @type {HTMLElement}
	 */
	let sidebar

	function toggle(site) {
		if (site.sidebar) {
			sidebar.classList.remove('-translate-x-full')
		} else {
			sidebar.classList.add('-translate-x-full')
		}
	}
	$: if (sidebar) toggle($site)
</script>

<aside
	bind:this={sidebar}
	class="flex flex-col flex-shrink-0 w-80 bg-skin-base py-4 absolute inset-y-0 left-0 -translate-x-full transform transition duration-200 ease-in-out z-5 border-r border-skin-subtle overflow-auto"
	class:lg:relative={$site.sidebar}
>
	<slot />
</aside>
