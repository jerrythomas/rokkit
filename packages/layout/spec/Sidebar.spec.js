import { describe, expect, beforeEach, it } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import Sidebar from '../src/Sidebar.svelte'

describe('Sidebar.svelte', () => {
	beforeEach(() => cleanup())

	it('should render sidebar', () => {
		const { container } = render(Sidebar, {
			items: [{ name: 'Alpha' }, { name: 'Beta' }]
		})
		expect(container).toMatchSnapshot()
	})
})
