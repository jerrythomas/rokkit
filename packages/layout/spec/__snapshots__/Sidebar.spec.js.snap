// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar.svelte > should render sidebar 1`] = `
<div>
  <aside
    class="sidebar flex flex-col"
  >
    <!---->
     
    <!---->
     
    <!---->
  </aside>
  
</div>
`;
